src/OpenTK.Mathematics/Vector/Vector4i.cs

```diff
@@ -0,0 +1,1633 @@
+        /// </summary>
+        public static readonly Vector4i Zero = new Vector4i(0, 0, 0, 0);
+
+        /// <summary>
+        /// Defines an instance with all components set to 1.
+        /// </summary>
+        public static readonly Vector4i One = new Vector4i(1, 1, 1, 1);
+
+        /// <summary>
+        /// Defines the size of the <see cref="Vector4i"/> struct in bytes.
+        /// </summary>
+        public static readonly int SizeInBytes = Marshal.SizeOf<Vector4i>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Vector4i"/> struct.
+        /// </summary>
+        /// <param name="value">The value that will initialize this instance.</param>
+        public Vector4i(int value)
+        {
+            X = value;
+            Y = value;
+            Z = value;
+            W = value;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Vector4i"/> struct.
+        /// </summary>
+        /// <param name="x">The X component of the Vector4i.</param>
+        /// <param name="y">The Y component of the Vector4i.</param>
+        /// <param name="z">The Z component of the Vector4i.</param>
+        /// <param name="w">The W component of the Vector4i.</param>
+        public Vector4i(int x, int y, int z, int w)
+        {
+            X = x;
+            Y = y;
+            Z = z;
+            W = w;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Vector4i"/> struct.
+        /// </summary>
+        /// <param name="v">The Vector2 to copy components from.</param>
+        public Vector4i(Vector2i v)
+        {
+            X = v.X;
+            Y = v.Y;
+            Z = 0;
+            W = 0;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Vector4i"/> struct.
+        /// </summary>
+        /// <param name="v">The <see cref="Vector3i"/> to copy components from.</param>
+        public Vector4i(Vector3i v)
+        {
+            X = v.X;
+            Y = v.Y;
+            Z = v.Z;
+            W = 0;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Vector4i"/> struct.
+        /// </summary>
+        /// <param name="v">The <see cref="Vector3i"/> to copy components from.</param>
+        /// <param name="w">The w component of the new Vector4.</param>
+        public Vector4i(Vector3i v, int w)
+        {
+            X = v.X;
+            Y = v.Y;
+            Z = v.Z;
+            W = w;
+        }
+
+        /// <summary>
+        /// Gets or sets the value at the index of the vector.
+        /// </summary>
+        /// <param name="index">The index of the component from the vector.</param>
+        /// <exception cref="IndexOutOfRangeException">Thrown if the index is less than 0 or greater than 3.</exception>
+        public int this[int index]
+        {
+            get
+            {
+                if (index == 0)
```

Comment:
Probably the floating point vectors do it the same way. But perhaps it would make sense to do it with FieldOffsets like this:
```
[StructLayout(Explicit)]
unsafe struct Vec4i
{
    [FieldOffset(0)]
    private fixed int _elements;
    [FieldOffset(0)]
    public int X;
    [FieldOffset(4)]
    public int Y;
    [FieldOffset(8)]
    public int Z;
    [FieldOffset(12)]
    public int W;
    public int this[int index]
    {
       get
       {
         if ((uint)index > 3) // faster check than (index > 3 || index < 0) at least on my machine -> should translate to jae(unsigned check) instruction
         {
           throw new IndexOutOfRangeException("You tried to access this vector at index: " + index);
         }
         return _elements[index];
       }
    }
}
```

Unfortunately I do not know if the fixed array is actually faster, or cause of some Jitter magic it get's slower, would need to test that still.
But it is a lot less branching, and therefore I at least would think it is.

Reply:
This was discussed earlier this year in the Discord server. While it was a speed boost (from ~4.7 ns to around ~0.7), a consequence of these "unions" are that the compiler doesn't know when all of the memory has been set. If we only set X, Y, Z, and W, it'll think `_elements` still needs to be set and require a `: this()` added to the constructors, as well as a constructor call. Methods like:

```cs
public static void Add(ref Vector3 a, ref Vector3 b, out Vector3 result)
 {
     result.X = a.X + b.X;
    result.Y = a.Y + b.Y;
    result.Z = a.Z + b.Z;
}       
```

would have to become:

```cs
public static void Add(ref Vector3 a, ref Vector3 b, out Vector3 result)
{
    result = new Vector3
    {
        X = a.X + b.X,
        Y = a.Y + b.Y,
        Z = a.Z + b.Z
    };
}
```

The consensus was that it's not worth trading the performance of these methods for the performance of the type conversions.

Reply:
Would it maybe make sense to use the int array internally, and have X, Y, Z, and W as wrappers around array[0], [1], [2], and [3]?

Reply:
Would need someone to test performance, because it it's a lot worse, I would prefer to leave it the old way, as I do not know a lot of use cases for the indexer on vectors...